### DISCUSSION TITLE
MLForecast Cross Validation: Utilizing Native Categorical Handling

### DISCUSSION THREAD

**Discussion by jrodenbergrheem** [Nixtla/mlforecast#421] [General]
I was curious as to how we can utilize native categorical handling of models like LGBM in mlforecast cross validation.

 I have seen the example of custom training using numpy arrays but unsure of how to use this with the mlf.cross_validation method. I have currently been specifying cateogrical indexs for lightgbm in the parameters argument but I get the message that "categorical_feature keyword will be ignored" 

Is this currently implemented and if so does anyone have a link to an example: this being done in polars is huge plus.

Thanks!

**Comment by jmoralez** [comment]
Hey. If you're using pandas you can just set their types as categorical and LightGBM will treat them accordingly. Are you using polars?

**Reply by jrodenbergrheem** [reply]
Ok great. I appreciate the quick response! Is native handling possible in polars?

**Reply by jmoralez** [reply]
LightGBM doesn't natively support polars categorical columns yet, but you can override the fit and predict methods to do something like what is done in the numpy guide:
```python
import polars as pl
from lightgbm import LGBMRegressor
from mlforecast import MLForecast
from utilsforecast.data import generate_series

class MyLGBM(LGBMRegressor):
    def _cat_to_int(self, X):
        return X.with_columns(*[pl.nth(c).cast(pl.Int64) for c in self.cat_cols])
    
    def fit(self, X, y):
        self.cat_cols = [i for i, dtype in enumerate(X.schema.dtypes()) if dtype == pl.Categorical]
        X = self._cat_to_int(X)
        return super().fit(X=X, y=y, categorical_feature=self.cat_cols)

    def predict(self, X):
        X = self._cat_to_int(X)
        return super().predict(X=X)

series = generate_series(5, engine='polars')
mlf = MLForecast(
    models=[MyLGBM(n_estimators=2)],
    freq='1d',
)
mlf.fit(series, static_features=['unique_id']).predict(2)
```
